src/compile/data/bin.ts

```diff
@@ -18,37 +18,41 @@ function numberFormatExpr(expr: string, format: string) {
 
 function parse(model: Model): Dict<VgTransform[]> {
   return model.reduceFieldDef(function(binComponent: Dict<VgTransform[]>, fieldDef: FieldDef, channel: Channel) {
-    const bin = model.fieldDef(channel).bin;
+    const fieldDefBin = model.fieldDef(channel).bin;
+    const bin: Bin = isBoolean(fieldDefBin) ? {} : fieldDefBin;
+    if (bin && !(<VgTransform>bin).maxbins && !(<VgTransform>bin).steps) {
+      (<VgTransform>bin).maxbins = autoMaxBins(channel);
+    }
+    // Make this bin variable always an object
+    // and extend it with maxbins if needed here.
+    const transform: VgTransform[] = [];
     if (bin) {
-
-      let binTrans: VgTransform = extend({
-        type: 'bin',
-        field: fieldDef.field,
-        as: [field(fieldDef, {binSuffix: 'start'}), field(fieldDef, {binSuffix: 'end'})]
-      },
-        // if bin is an object, load parameter here!
-        typeof bin === 'boolean' ? {} : bin
-      );
-
-      const transform: VgTransform[] = [];
-      if (!binTrans.extent) {
-        const extentSignal = model.getName(fieldDef.field + '_extent');
-        transform.push({
-          type: 'extent',
-          field: fieldDef.field,
-          signal: extentSignal
-        });
-
-        binTrans.extent = {signal: extentSignal};
+      const key = hash(bin) + '_' + fieldDef.field;
+      if (binComponent[key]) {
+        transform.push.apply(transform, binComponent[key]); // concat in place
+      } else {
+        const checkExtentAbsent: boolean = !bin.extent;
```

Comment:
This variable makes it harder to read the code, just simply remove it and use `!bin.extent`